### PR TITLE
Draft of dependency graph generation algorithm

### PR DESCRIPTION
**Pull request by atgeller**
### Description

This is a draft of an analysis to determine dependency information between qubit operations in a quake program.

### CODE REVIEW

**Comment by atgeller** [comment]
I have read the Contributor License Agreement and I hereby accept the Terms.

File: lib/Optimizer/Transforms/DependencyAnalysis.cpp [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    }
+  };
+
+  DependencyNode(DependencyNode *from)
+      : successors({from}), dependencies(), qindex(from->qindex),
+        associated(nullptr) {
+    addSuccessor(from);
+  };
+
+  inline void addSuccessor(DependencyNode *other) {
```
Comment (schweitzpgi) [inline lib/Optimizer/Transforms/DependencyAnalysis.cpp:69]: nit: the `inline` is redundant here

**Review by schweitzpgi — COMMENTED** [review]
I'll have to try this out.

**Comment by atgeller** [comment]
Can be used like so: `cudaq-quake FILE | cudaq-opt --add-dealloc --factor-quantum-alloc --memtoreg --cse --canonicalize --assign-ids --dep-analysis`. Use `--debug` to see lifetime info printed. An example test case can be found in `test/Quake/dependency.qke`.

The additional `assign-ids` pass is a necessary first step to assign matching unique identifiers to each qubit alloc/dealloc.

The `dep-analysis` pass builds dependency graphs for straightline code, performs scheduling, determines lifetimes and finds reuse opportunities, and then emits a new program based on the new schedule and reuses.

File: include/cudaq/Optimizer/Dialect/Quake/QuakeOps.h [hunk 2]
```diff
@@ -70,6 +70,15 @@ void genericOpPrinter(mlir::OpAsmPrinter &_odsPrinter, mlir::Operation *op,
 // Utility functions to test the form of an operation.
 //===----------------------------------------------------------------------===//
 
+// Is \p op in the Quake dialect?
+// TODO: Is this StringRef comparison faster than calling MLIRContext::
+// getLoadedDialect("quake")?
```
Comment (schweitzpgi) [inline include/cudaq/Optimizer/Dialect/Quake/QuakeOps.h:75]: I believe it must be. This does a O(c) —where c is at most 5—comparison on two character sequences. The getLoadedDialect call does an O(log_2(n)) search over all loaded dialects with the same constant factors.

File: lib/Optimizer/Transforms/AssignIDs.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+  else if (hasClassicalInput(defop))
+    i++;
+
+  return findQid(defop->getOperand(i));
+}
+
+class SinkOpPat : public OpRewritePattern<quake::SinkOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  using Base = OpRewritePattern<quake::SinkOp>;
```
Comment (schweitzpgi) [inline lib/Optimizer/Transforms/AssignIDs.cpp:110]: nit: can this line be deleted?

File: lib/Optimizer/Transforms/AssignIDs.cpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    auto *ctx = &getContext();
+    func::FuncOp func = getOperation();
+    RewritePatternSet patterns(ctx);
+    unsigned x = 0;
+    patterns.insert<NullWirePat>(ctx, &x);
+    patterns.insert<SinkOpPat>(ctx);
+    ConversionTarget target(*ctx);
+    target.addLegalDialect<quake::QuakeDialect>();
+    target.addDynamicallyLegalOp<quake::NullWireOp>(
+        [&](quake::NullWireOp alloc) { return alloc->hasAttr("qid"); });
```
Comment (schweitzpgi) [inline lib/Optimizer/Transforms/AssignIDs.cpp:160]: The attribute approach is brittle as other transformations will not necessarily keep the information. We should really use WireSet, BorrowWire, ReturnWire for explicitly mapping the identity of a qubit from an infinite space (virtual registers) to a finite space (physical registers).

Comment (atgeller) [reply lib/Optimizer/Transforms/AssignIDs.cpp:160]: Makes sense, right now Assign IDs and Dependency Analysis are very tangled passes, so they pretty much have to be run one right after another, and Dependency Analysis has to check that the attributes are present.

File: lib/Optimizer/Transforms/AssignIDs.cpp (hunk 3 above)
Comment (atgeller) [reply lib/Optimizer/Transforms/AssignIDs.cpp]: Yes, done

File: include/cudaq/Optimizer/Dialect/Quake/QuakeOps.h (hunk 2 above)
Comment (atgeller) [reply include/cudaq/Optimizer/Dialect/Quake/QuakeOps.h:75]: Gotcha, I brought the comment over with the rest of the function so I didn't know.

**Comment by atgeller** [comment]
Closing in lieu of #2163

**Pull request closed without merging**